File: go/problems/day1.go
package problems

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func convertNumber(str string) uint8 {
	nums := []string{
		"one", "two", "three", "four", "five",
		"six", "seven", "eight", "nine",
	}

	for i, v := range nums {
		if strings.HasPrefix(str, v) {
			return uint8(i) + 1
		}
	}

	return 0
}

func getNumber(str string) int {
	var firstDigit, lastDigit uint8
	for i := 0; i < len(str); i++ {
		if str[i] > 47 && str[i] < 58 {
			firstDigit = str[i] - '0'
			break
		}

		if n := convertNumber(str[i:]); n > 0 {
			firstDigit = n
			break
		}
	}
	for i := len(str) - 1; i >= 0; i-- {
		if str[i] > 47 && str[i] < 58 {
			lastDigit = str[i] - '0'
			break
		}

		if n := convertNumber(str[i:]); n > 0 {
			lastDigit = n
			break
		}
	}
	num := firstDigit*10 + lastDigit
	return int(num)
}

func Day1() {
	file, err := os.Open("../input/day1.txt")
	defer file.Close()

	if err != nil {
		fmt.Println(err)
	}
	fileScanner := bufio.NewScanner(file)
	fileScanner.Split(bufio.ScanLines)

	sum := 0
	for fileScanner.Scan() {
		line := fileScanner.Text()
		num := getNumber(line)
		sum += num
		fmt.Println(num, line)
	}
	fmt.Println(sum)
}

File: rust/src/problems/day1.rs
use std::fs;

pub fn solve() {
    let sum: i32 = fs::read_to_string("../input/day1.txt".to_string())
        .unwrap()
        .lines()
        .map(transform)
        .sum();
    println!("day1: {sum}")
}

fn transform(line: &str) -> i32 {
    let mut c1 = '0';
    let mut c2 = '0';
    for i in 0..line.len() {
        let c = line.chars().nth(i).unwrap();
        if c.is_digit(10) {
            c1 = c;
            break;
        }
        c1 = convert(&line[i..]);
        if c1 != '0' {
            break;
        }
    }
    for i in (0..line.len()).rev() {
        let c = line.chars().nth(i).unwrap();
        if c.is_digit(10) {
            c2 = c;
            break;
        }
        c2 = convert(&line[i..]);
        if c2 != '0' {
            break;
        }
    }
    format!("{}{}", c1, c2).parse().unwrap()
}

fn convert(line: &str) -> char {
    match line {
        ln if ln.starts_with("one") => '1',
        ln if ln.starts_with("two") => '2',
        ln if ln.starts_with("three") => '3',
        ln if ln.starts_with("four") => '4',
        ln if ln.starts_with("five") => '5',
        ln if ln.starts_with("six") => '6',
        ln if ln.starts_with("seven") => '7',
        ln if ln.starts_with("eight") => '8',
        ln if ln.starts_with("nine") => '9',
        _ => '0',
    }
}
